pkg/config/generate.go

```diff
@@ -388,3 +394,40 @@ func Generate(cs *api.OpenShiftManagedCluster) (err error) {
 
 	return
 }
+
+// regenerate checks if certificate needs to be regenerated
+func regenerate(cs, oldCs *api.OpenShiftManagedCluster, cert certificate) bool {
```

Comment:
this function needs to check the inputs against the pre-existing certificate in cs.Config to determine if a regeneration is needed.

Reply:
This is where the challenge comes. My certificate might be for `old.example.com` but my `cs` is coming from `newawesomedomain.com`. So without `oldCs` im not able to check which certificate depends to where because values might be totally different. 
Alternatives are to keep some kind dependency between certificates by adding `type` to the certificate and mapping `type=masterFrontEnd` needs regeneration if `PublicHostname` or `FQDN` changes.

Reply:
If you can think of any other way, let me know.

Reply:
Is VerifyHostname what you need here?

https://golang.org/pkg/crypto/x509/#Certificate.VerifyHostname

Reply:
The thing is I dont have a reference point to what to compare it. Without oldCS I need to know which cert uses which Properties field. In the example, routerPublicHostname is very similar to API FQDN, but they are not related. So verifyHostname will not give me this relationship.

Reply:
At this point all the hostnames in cs are updated to new, so if a certificate does not match **any** hostname then you probably want to regenerate it, right?

Reply:
This is just one case. In example:
```
 {
            cn: cs.Properties.RouterProfiles[0].PublicSubdomain,
            dnsNames: []string{
                cs.Properties.RouterProfiles[0].PublicSubdomain,
                "*." + cs.Properties.RouterProfiles[0].PublicSubdomain,
            },
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.Router.Key,
            cert:        &c.Certificates.Router.Cert,
        },
        {
            cn: "docker-registry-default." + cs.Properties.RouterProfiles[0].PublicSubdomain,
            dnsNames: []string{
                "docker-registry-default." + cs.Properties.RouterProfiles[0].PublicSubdomain,
                "docker-registry.default.svc",
                "docker-registry.default.svc.cluster.local",
            },
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.Registry.Key,
            cert:        &c.Certificates.Registry.Cert,
        },
        {
            cn: cs.Properties.FQDN,
            dnsNames: []string{
                cs.Properties.FQDN,
                "master-000000",
                "master-000001",
                "master-000002",
                "kubernetes",
                "kubernetes.default",
                "kubernetes.default.svc",
                "kubernetes.default.svc.cluster.local",
            },
            ipAddresses: []net.IP{net.ParseIP("172.30.0.1")},
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.MasterServer.Key,
            cert:        &c.Certificates.MasterServer.Cert,
            
        }, 
```
certificates does not have purpose/type. Just fields If I change Router FQDN to totally different value, how do I know that I need to regenerate router certificate, and not Master or Service Catalog? They are not related, but without mapping keeping somewhere just matching is not enough.

Reply:
Because master and service catalog are still going to match with the other hostname.

Reply:
So you suggest to take app properties values, iterate over small subset of certificates (the ones which depend on those properties) and if none of them matches properties field, regenerate it. If match found - skip

Reply:
I didn't suggest that but it's also an option to group certificates per "parent" hostname. I suggested that given a certificate, you can run VerifyHostname on it against all available hostnames (publicHostname, routerPublicHostname). If the cert cannot verify against any hostname, you want to regenerate it.

Reply:
@mjudeikis you need to write a function which compares the current certificate's fields to the cn, organization, dnsNames, ipAddresses, extKeyUsage desired state.  If they don't match, regenerate.  We already do this to determine if the CA has changed and we need to regenerate, so what's left is the easier bit?